### DISCUSSION TITLE
Emacs Compatibility Tracking Issue

### DISCUSSION THREAD

**Discussion by morgoth** [zed-industries/zed#23629] [Feature Requests]
Rather than a monolithic "implement emacs in Zed" enhancement request (which is inherently unactionable) this issue is a tracking issue for centralizing various Zed bug fixes and enhancements to better supporting emacs-like compatibility and operation in Zed. 

Issues:
- https://github.com/zed-industries/zed/issues/4697
- https://github.com/zed-industries/zed/issues/8580
- https://github.com/zed-industries/zed/issues/8614 
- https://github.com/zed-industries/zed/issues/9230 
- https://github.com/zed-industries/zed/issues/10910
- https://github.com/zed-industries/zed/issues/14260
- https://github.com/zed-industries/zed/issues/16214
- https://github.com/zed-industries/zed/issues/18420
- https://github.com/zed-industries/zed/issues/20986
- https://github.com/zed-industries/zed/issues/22490
- https://github.com/zed-industries/zed/issues/22943
- https://github.com/zed-industries/zed/issues/23408

Merged PRs:
- https://github.com/zed-industries/zed/pull/19605
- https://github.com/zed-industries/zed/pull/21003
- https://github.com/zed-industries/zed/pull/21921
- https://github.com/zed-industries/zed/pull/21923
- https://github.com/zed-industries/zed/pull/22590
- https://github.com/zed-industries/zed/pull/22629
- https://github.com/zed-industries/zed/pull/22779
- https://github.com/zed-industries/zed/pull/22904
- https://github.com/zed-industries/zed/pull/22974
- https://github.com/zed-industries/zed/pull/23297
- https://github.com/zed-industries/zed/pull/23428
- https://github.com/zed-industries/zed/pull/25284
- https://github.com/zed-industries/zed/pull/30858

Comments such as "I need emacs mode to switch to Zed!" and the like are unwelcome here.
Please vote with 👍 below on this issue and the linked issues instead. 

Editor note (@notpeter): Originally this issue was for an emacs keymap, a beta of which shipped in https://github.com/zed-industries/zed/pull/19605

**Comment by ozanvos** [comment]
An offical "Emacs mode" would be really nice. There is a couple of example user configurations over here: https://github.com/zed-industries/community/discussions/1653

At the moment the lack of a "mark" command is the biggest holdup. In Emacs, pressing <kbd>Ctrl+SPC</kbd> will toggle selection and let you use movement commands to select text without keeping <kbd>Shift</kbd> pressed. Any chance this could be added to Zed?

**Comment by lvii** [comment]
+1

**Comment by yonkeltron** [comment]
This is more or less what's stopping me from adopting Zed.

**Comment by lllleonnnn** [comment]
Yes please...

**Comment by opsnull** [comment]
Please add Emacs keymap supporting.

**Comment by DidacBable** [comment]
I would also appreciate this a lot. This is a standard feature of other IDEs like VSCode and Intellij's IDEs.

**Comment by samin-UMG** [comment]
Stopping me too. I'm too used to emacs keybindings. There is no way at this point I can unlearn them.

**Comment by djhunter67** [comment]
I would love to use this editor on my Arch box, btw, but I am so accustomed to Emacs.  Is there any information regarding and movement on this issue?

**Comment by akaihola** [comment]
>At the moment the lack of a "mark" command is the biggest holdup. In Emacs, pressing `Ctrl+SPC` will toggle selection and let you use movement commands to select text without keeping `Shift` pressed. Any chance this could be added to Zed?

At least there's an issue for this (go and vote!):
- #8580

**Comment by notpeter** [comment]
For those following this issue we've recently shipped two emacs-related improvements:
- https://github.com/zed-industries/zed/pull/19605
- https://github.com/zed-industries/zed/pull/21003

Please try out the `Emacs (beta)` keymap by triggering the `welcome: toggle bas keymap selector`: 

![image](https://github.com/user-attachments/assets/bc9dbe88-afab-422e-bb39-89a508594079)

It's very basic at this point, but give it a whirl, happy to consider PRs with improvements / refinements.

Rather than close this issue I've repurposed it as a tracking issue for additional Emacs-related enhancements and I've edited the issue description above to include links to other issues.

One other update, sadly, due to issues with international keyboards we also opted to revert `terminal.option_as_meta` to default to false again:
- https://github.com/zed-industries/zed/pull/19364

If you need emacs/readline style keybinds in your terminal which use `meta` / `alt` you will want to add this to your settings.json:
```json
  "terminal": {
    "option_as_meta": true
  }
```

**Comment by jinschoi** [comment]
Some additional keybindings that I very frequently use and aren't in the default:

```json
[
  {
    "context": "Editor",
    "bindings": {
      "alt-<": "editor::MoveToBeginning",
      "alt->": "editor::MoveToEnd",
      "alt-u": "editor::ConvertToUpperCase",
      "alt-l": "editor::ConvertToLowerCase",
      "alt-c": "editor::ConvertToUpperCamelCase"
    }
  },
  {
    "context": "Editor && mode == full",
    "bindings": {
      "ctrl-o": ["workspace::SendKeystrokes", "enter ctrl-p ctrl-e"]
    }
  },
  {
    "context": "BufferSearchBar > Editor",
    "bindings": {
      "ctrl-s": "search::SelectNextMatch",
      "ctrl-r": "search::SelectPrevMatch",
      "ctrl-g": "buffer_search::Dismiss"
    }
  }
]
```

First section is alternative bindings for MoveToBeginning and MoveToEnd, and case conversion for the word after the point.

Second section is for ^o, which is very often used to add a newline at point without moving the cursor. It is _not_ the same as "editor::NewlineBelow" which adds a newline below the current line and moves the cursor to it. Sending keystrokes isn't perfect since typing enter will sometimes add characters depending on context, for example if you are in a comment it may add a comment prefix to the next line, but it is close enough.

The last section is for repeated forward and backward searches after a search has been initiated. There should also be a ctrl-r keybinding in the Editor context to initiate a search backwards, but "buffer_search::Deploy" always begins by selecting the next hit and it doesn't appear to be any way to have it do that.

**Comment by djhunter67** [comment]
> Some additional keybindings that I very frequently use and aren't in the default:
> 
> ```json
> [
>   {
>     "context": "Editor",
>     "bindings": {
>       "alt-<": "editor::MoveToBeginning",
>       "alt->": "editor::MoveToEnd",
>       "alt-u": "editor::ConvertToUpperCase",
>       "alt-l": "editor::ConvertToLowerCase",
>       "alt-c": "editor::ConvertToUpperCamelCase"
>     }
>   },
>   {
>     "context": "Editor && mode == full",
>     "bindings": {
>       "ctrl-o": ["workspace::SendKeystrokes", "enter ctrl-p ctrl-e"]
>     }
>   },
>   {
>     "context": "BufferSearchBar > Editor",
>     "bindings": {
>       "ctrl-s": "search::SelectNextMatch",
>       "ctrl-r": "search::SelectPrevMatch",
>       "ctrl-g": "buffer_search::Dismiss"
>     }
>   }
> ]
> ```
> 
> First section is alternative bindings for MoveToBeginning and MoveToEnd, and case conversion for the word after the point.
> 
> Second section is for ^o, which is very often used to add a newline at point without moving the cursor. It is _not_ the same as "editor::NewlineBelow" which adds a newline below the current line and moves the cursor to it. Sending keystrokes isn't perfect since typing enter will sometimes add characters depending on context, for example if you are in a comment it may add a comment prefix to the next line, but it is close enough.
> 
> The last section is for repeated forward and backward searches after a search has been initiated. There should also be a ctrl-r keybinding in the Editor context to initiate a search backwards, but "buffer_search::Deploy" always begins by selecting the next hit and it doesn't appear to be any way to have it do that.
> 

This is spectacular!  Thank you so much!!

**Comment by pbgc** [comment]
I added some additional bindings that I use are also are not in the default:

```
[
  {
    "bindings": {
      "ctrl-g": "menu::Cancel"
    }
  },
  {
    "context": "Editor",
    "bindings": {
      "alt-<": "editor::MoveToBeginning",
      "alt->": "editor::MoveToEnd",
      "alt-u": "editor::ConvertToUpperCase",
      "alt-l": "editor::ConvertToLowerCase",
      "alt-c": "editor::ConvertToUpperCamelCase",
      "alt-.": "editor::GoToDefinition",
      "ctrl-x h": "editor::SelectAll"
    }
  },
  {
    "context": "Editor && mode == full",
    "bindings": {
      "ctrl-o": ["workspace::SendKeystrokes", "enter ctrl-p ctrl-e"],
      "alt-g alt-g": "go_to_line::Toggle"
    }
  },
  {
    "context": "BufferSearchBar > Editor",
    "bindings": {
      "ctrl-s": "search::SelectNextMatch",
      "ctrl-r": "search::SelectPrevMatch",
      "ctrl-g": "buffer_search::Dismiss",
      "enter": "buffer_search::Dismiss"
    }
  },
  {
    "context": "Pane",
    "bindings": {
      "ctrl-x 0": "pane::CloseActiveItem",
      "ctrl-x 1": "workspace::CloseInactiveTabsAndPanes",
      "ctrl-x 2": "pane::SplitDown",
      "ctrl-x 3": "pane::SplitRight",
      "alt-,": "pane::GoBack"
    }
  },
  {
    "context": "Workspace",
    "bindings": {
      "alt-x": "command_palette::Toggle"
    }
  }
]
```

**Comment by pasunboneleve** [comment]
Ideally we would have a wiki page or documentation for all the default keybindings that would make the experience in **Zed** most like **Emacs**. I already found some that I just can't live without, and this is bound to continue to happen:

`"context": "Workspace"`
```json      
      "ctrl-x 4": "workspace::NewWindow",
```

`"context": "Editor:"`
```json
    "alt-;": "editor::ToggleComments",
    "ctrl-/": "editor::Undo",
    "ctrl-shift-/": "editor::Redo"
```

**Comment by rakete** [comment]
The big thing missing for me is the transient mark mode: enabling marking text with ctrl+space, moving around while the mark stays active as if shift is being held, and then either aborting with ctrl+g or copy/cut with alt/ctrl+w.

I've worked around this with an ahk v2 script for now: 

```
#HotIf WinActive("ahk_class Zed::Window")

enabled := false
; ctrl+space enables transient mark mode by holding down shift
$^space::{
    global enabled := true
    SendInput('{LShift Down}')
}
; pressing ctrl+space again while shift is held down disables
; transient mark mode and releases shift
$+^space::{
    if enabled {
        global enabled := false
        SendInput('{LShift Up}')
        ; this "wiggles" the cursor once to get rid of the active selection
        SendInput('{Esc}{Up}{Down}')
    } else {
        ; if it wasn't enabled, send ctrl+shift+space
        SendInput('+^space')
    }
}

; ctrl+w (and shift) while transient mark mode is enabled
; cuts text with ctrl+x and disables transient mark mode
$+^w::{
    if enabled {
        SendInput('{LShift Up}')
    }
    SendInput('^x')
    if enabled {
        global enabled := false
        SendInput('{Esc}{Up}{Down}')
    }
}
; alt+w (and shift) while transient mark mode is enabled
; copies text with ctrl+c and disables transient mark mode
$+!w::{
    if enabled {
        SendInput('{LShift Up}')
    }
    SendInput('^c')
    if enabled {
        global enabled := false
        SendInput('{Esc}{Up}{Down}')
    }
}
; ctrl+y pastes text with shift+insert
$*^y::{
    SendInput('+{Ins}')
}

; ctrl+g disables transient mark mode
$*^g::{
    if enabled {
        global enabled := false
        SendInput('{LShift Up}')
        ; escape clears the selection
        SendInput('{Esc}{Up}{Down}')
    } else {
        SendInput('{Esc}')
        ; if it wasn't enabled, send ctrl+g
        SendInput('^g')
    }
}

; esc (and shift) disables transient mark mode
$+esc:: {
    if enabled {
        global enabled := false
        SendInput('{LShift Up}')
        ; escape clears the selection
        SendInput('{Esc}{Up}{Down}')
    } else {
        ; if it wasn't enabled, send shift+esc
        SendInput('+{Esc}')
    }
}

#HotIf
```

**Comment by notpeter** [comment]
I've added a bunch of the suggestions provided above and hidden the associated comments to keep this thread from growing out of control.  Where possible I'm trying to stick only the actual default emacs keybinds and avoid people's personal binds.

Here are some recent improvements which have made their way into Zed:

- #21921 is available in v0.167.x (since last Wednesday 2025-01-01 in Zed Stable)
- #22590 will be available in v0.169.x (from Wednesday in Zed Preview; Zed Stable 2025-01-15)

@rakete Nifty that There's you've managed to build a Mark mode workaround with AutoHotKey.
As mentioned above, there's also a a dedicated issue:
- https://github.com/zed-industries/zed/issues/8580

**Comment by pasunboneleve** [comment]
I'm a big fan of **occur**. Would it be too much to add it to a task? Obviously that's functionality that goes well beyond keybindings.

**Comment by notpeter** [comment]
@dmvianna I don't have any experience using emacs `occur` but fundamentally it looks quite similar to a Zed Project Search multibuffer. Zed multibuffers support editing across multiple files while paging through search results.  Currently multibuffers default to +2 / -2 lines of context above/below the line in question, I imagine it would be possible to create an `occur` search multibuffer that have 0 lines of context and potentially hide 'expand context up/down' buttons to create a more minimalist view which more closely resembles occur.  Feel free to file an issue. To be upfront, it's not something we're likely to immediately prioritize, but it definitely would be worth tracking user interest.

Edit: Thanks for filing an issue:
- https://github.com/zed-industries/zed/issues/22943

**Comment by kzar** [comment]
It might be worth adding `"ctrl-o": "editor::Newline"` for the editor context?

**Comment by notpeter** [comment]
@kzar Emacs `ctrl-o` behavior is slightly different that `editor::Newline`, `ctrl-o` creates a newline below but does not move the cursor.  I would prefer to add support for an optional parameter to the `editor::NewlineBelow` action so that we could bind `"ctrl-o": ["editor::NewlineBelow", {"move_cursor": false}]`.

Similar to this:
https://github.com/zed-industries/zed/blob/ba7b9b33c9ca0d0899583b93c3636ea7e5849c01/assets/keymaps/default-linux.json#L306
https://github.com/zed-industries/zed/blob/ba7b9b33c9ca0d0899583b93c3636ea7e5849c01/crates/editor/src/actions.rs#L79-L85
And then add a conditional inside the `newline_below` to optionally move the cursor. 
https://github.com/zed-industries/zed/blob/ba7b9b33c9ca0d0899583b93c3636ea7e5849c01/crates/editor/src/editor.rs#L3139

**Comment by kzar** [comment]
Good point about not moving the cursor, but I don't think that `editor::NewlineBelow` is right because if you're at the start of a line C-o should move the current line down (or if you're in the middle of a line, it should move the part after your cursor down). Perhaps we could go with your approach to stop the cursor moving, but with `editor:Newline`?

**Comment by kzar** [comment]
`"ctrl-t": "editor::Transpose"` for the editor context is another one I kept missing. _(@notpeter is it helpful if I keep commenting with these, or if not is there a better way to share them that would help?)_

**Comment by pasunboneleve** [comment]
> @dmvianna I don't have any experience using emacs `occur` but fundamentally it looks quite similar to a Zed Project Search multibuffer. 

I opened a feature request, where I delve more deeply into the differences between `occur` and `Zed Project Search Multibuffer`: https://github.com/zed-industries/zed/issues/22943

**Comment by pasunboneleve** [comment]
Oh, duh.

`"context": "Editor:"`
```json
    "ctrl-y": "editor::Paste",
    "alt-w": "editor::Copy",
    "ctrl-w": "editor::Cut",
    "ctrl-?": "editor::Redo",
    "ctrl-shift-n": "editor::SelectDown",
    "ctrl-shift-p": "editor::SelectUp",
    "alt-shift-f": "editor::SelectToNextWordEnd",
    "alt-shift-b": "editor::SelectToPreviousWordStart",
    "ctrl-alt-shift-f": "editor::SelectToEndOfLine",
    "ctrl-alt-shift-b": "editor::SelectToBeginningOfLine"
```

This below also stops the `Assistant` from opening at every attempt to run `editor::Redo`.

`"bindings"`
```json
  "alt-x": "app_menu::OpenApplicationMenu",
  "ctrl-?": null
  ```

And this prevents the cursor movement commands from opening new buffers when reaching beyond the first and last lines of a buffer. 

`"context": "Workspace:"`
```json
  "ctrl-n": null,
  "ctrl-p": null,
```

**Comment by pasunboneleve** [comment]
Another biggie: the [kill ring](https://www.gnu.org/software/emacs/manual/html_node/emacs/Yanking.html). Basically a copy history that you can browse through by typing `alt-y` after the first `ctrl-y` (paste). And all its associated commands, as actually seeing the list of copy items in memory before choosing one (pressing `alt-y` not preceded by `ctrl-y` or itself).

There's an [issue](https://github.com/zed-industries/zed/issues/22490) for it already, it would be nice to have it in the description of this issue too.

**Comment by pasunboneleve** [comment]
Moving around the workspace panes is different from **Emacs**, which will just cycle through all available buffers and has no default concept of how they're positioned. **Zed** seems to organise them spatially, so instead of having just the default **Emacs** `ctrl-x o` (which I rebinded to `alt-o` as per [Mastering Emacs](https://www.masteringemacs.org/) in my own configuration), I had to use spatial bindings. I dislike it; I find **Emacs**'s approach better, especially when I can use **IBuffer** to locate buffers by name using `ctrl-x b`.

In **Emacs**, I go to the pane I want the buffer to be and switch to it by name. In **Zed** (at least for now), I have to locate the pane by looking at the tabs, switching to the pane and then choosing the desired tab. Suboptimal.

Here's my current config:

`"context": "Workspace"`
```json
  "ctrl-x o": "tab_switcher::Toggle",
  "alt-x b": ["workspace::ActivatePaneInDirection", "Left"],
  "alt-x f": ["workspace::ActivatePaneInDirection", "Right"],
  "alt-x p": ["workspace::ActivatePaneInDirection", "Up"],
  "alt-x n": ["workspace::ActivatePaneInDirection", "Down"]
```

**Comment by pasunboneleve** [comment]
`pop-global-mark` in **Emacs** takes you back to the last mark, which in **Zed** it is the same as `xref-go-back`. 
`"context": "Pane"`
```json
  "alt-,": "pane::GoBack",  // xref-go-back
  "ctrl-x ctrl-space": "pane::GoBack",  // pop-global-mark
  "ctrl-u ctrl-space": "pane::GoForward"  // set-mark-command with negative prefix
```

Typing `ctrl-x ctrl-space` fast enough in **Zed** is tricky. How can I make **Zed** less impatient?

**Comment by failable** [comment]
Maybe related emacs issues: 
- https://github.com/zed-industries/zed/issues/4697
- https://github.com/zed-industries/zed/issues/9230 
- https://github.com/zed-industries/zed/issues/8614

**Comment by notpeter** [comment]
> At the moment the lack of a "mark" command is the biggest holdup. In Emacs, pressing Ctrl+SPC will toggle selection and let you use movement commands to select text without keeping Shift pressed. Any chance this could be added to Zed? -@ozanmakes [comment link](https://github.com/zed-industries/zed/issues/4856#issuecomment-1908635833)

Tomorrow (2025-01-21) this will be available Zed Preview 0.171.x (https://github.com/zed-industries/zed/pull/23297 and https://github.com/zed-industries/zed/pull/23428):

- `ctrl-space` to enter mark/selection mode (`editor::SetMark`)
- `ctrl-x ctrl-x` to swap selection ends (`editor::SwapSelectionEnds` aka `exchange-point-and-mark`)

Note multi-cursor is supported for both.

<details><summary>video demo</summary>

https://github.com/user-attachments/assets/d4104965-369d-4e54-97c9-4c8f9a42eb36

</details>

Thanks @ozanmakes for your work on https://github.com/zed-industries/zed/pull/22904 which inspired me to get this over the line.

**Comment by kzar** [comment]
Zed is getting more and more usable for me (as a long-time Emacs user) with these changes, thanks @notpeter!

I've not had the latest upgrade yet, but something I've noticed recently. If I press `ctrl-k` (editor: kill ring cut) to delete to the end of the line, then `ctrl-y` (kill ring yank) the text will be pasted back as expected. But if I then select some other text, press `ctrl-w` (editor: cut), then `ctrl-y`, the wrong text will be pasted (the text from the original king ring cut). In Emacs, I can paste back the text after using `ctrl-w`.

**Comment by michaelgruner** [comment]
> Tomorrow (2025-01-21) this will be available Zed Preview 0.171.x (https://github.com/zed-industries/zed/pull/23297 and https://github.com/zed-industries/zed/pull/23428):

Thanks for this! Zed is incredibly fast. This was one of my stoppers to finally switch over.

**Comment by maximvl** [comment]
I'm trying this Emacs keymap, one thing I'm confused about is how it should handle conflicts with default Zed keybinds, eg C-p instead of going to previous line currently opens project file search and C-a selects all instead of going to beginning of a line.

**Comment by michaelgruner** [comment]
> I'm trying this Emacs keymap, one thing I'm confused about is how it should handle conflicts with default Zed keybinds, eg C-p instead of going to previous line currently opens project file search and C-a selects all instead of going to beginning of a line.

You're probably missing the `base_keymap` in your `settings.json`?

```
"base_keymap": "Emacs"
```

**Comment by maximvl** [comment]
> You're probably missing the `base_keymap` in your `settings.json`?
> 
> ```
> "base_keymap": "Emacs"
> ```

I do have that, and I have some Emacs keybinds working, like C-n, C-e, but only those which are not used by Zed default bindings

**Comment by michaelgruner** [comment]
> I do have that, and I have some Emacs keybinds working, like C-n, C-e, but only those which are not used by Zed default bindings

`ctrl-p` works just fine for me, as well as all the other navigation keys, yank, paste, save, search, etc... Using the default base_keymap (no custom key binds). The only thing missing for me is setting a mark with `ctrl-space`.  

I'm in Zed `0.170.1 4b2e556bf3ff1c6cf98804cf6c855dd7776a4d1` on Mac. Here's my barebones `~/config/zed/settings.json`:

```
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run the `open default settings` command
// from the command palette or from `Zed` application menu.
{
  "assistant": {
    "default_model": {
      "provider": "copilot_chat",
      "model": "gpt-4o"
    },
    "version": "2"
  },
  "vim_mode": false,
  "theme": "Solarized Dark",
  "ui_font_size": 16,
  "buffer_font_size": 16,
  "autosave": {
    "after_delay": {
      "milliseconds": 10000
    }
  },
  "hour_format": "hour24",
  "tab_size": 2,
  "base_keymap": "Emacs",
  "terminal": {
    "option_as_meta": true
  }
}
```

And my `~/.config/zed/keymap.json` is empty.

**Comment by failable** [comment]
In Emacs, one can bind the same keyboard shortcut to different commands in different programming language buffers. It would be great if Zed had similar functionality.

For example, using the same shortcut to execute different run commands / tasks in Python and Rust files, e.g. `python3 $ZED_FILE` and `cargo run` respectively.

**Comment by pasunboneleve** [comment]
> To help Emacs folks migrate, can I suggest adding feedback for when the user used the command palette for a command that already has a key binding. That way, folks can pick up the keybindings as they go. I've been finding I need to use the command pallete to find what I need (e.g. "file finder: toggle"), then go through the default keymap myself to find what that's bound to.
> In Emacs, a message like You can run the command ‘forward-char’ with C-f is shown in the mini buffer, so the user can quickly pick up the binding for next time.

Second this. I used the dropdown boxes to discover command names and current keybindings in Zed, but a `which-key` functionality where the keybindings are shown alongside the commands in the pallette would be much more useful as it would cover all commands and be readily accessible.-- Daniel Vianna

**Comment by notpeter** [comment]
I've moved this to a discussion because we're working to tidy up the Issue Tracker.  We'll see how it goes.

Re which-key:
- https://github.com/zed-industries/zed/issues/10910

**Comment by EmasXP** [comment]
What do you think about support for s-expressions ("sexp" for short), like `Ctrl-Alt-Left` for `backward-sexp` and `Ctrl-Alt-Right` for `forward-sexp`?

Link to the GNU Emacs Manual: https://www.gnu.org/software/emacs/manual/html_node/emacs/Expressions.html

**Comment by michaelgruner** [comment]
Hey @notpeter I've updated Zed to 0.171.3 and I can confirm `ctrl-space` works as intended to set selection marks. A few improvements to feel more emacs-like would be:
- Copying with `alt-w` should remove the selection after copying
- Yanking (pasting) with `ctrl-y` is not working in the base keymap. I had to manually add it to my settings.

Thanks!

**Reply by pbgc** [reply]
@michaelgruner for now I defined alt-w as:

```
{
    "context": "Editor",
    "bindings": {
      "alt-w": ["workspace::SendKeystrokes", "cmd-c ctrl-g"],
    .....
``` 
so the selection is removed after copying.

I also had to define ctrl-y in my keymap (it was working before)

**Reply by michaelgruner** [reply]
Thanks, since we are sharing, I also did your trick for `cut` and mapped tab to auto-indent, which emacs had spoiled me already:
```
            "tab": "editor::AutoIndent",
            "alt-w": ["workspace::SendKeystrokes", "cmd-c ctrl-g"],
```

**Reply by pbgc** [reply]
@michaelgruner thanks! the tab mapped to auto-indent was one of the things I really was missing from emacs! I defined it like this to be able to keep using tab to auto-complete the ai sugestions:

```
  {
    "context": "Editor && !inline_completion && !showing_completions",
    "use_key_equivalents": true,
    "bindings": {
      "tab": "editor::AutoIndent"
    }
  },

```

**Reply by devcexx** [reply]
Hes, I've also felt that we're missing something feature for "exiting from the selection mode if something happens" (yes, probably the "something happens" is not pretty well defined). Even though the workspace::SendKeystrokes sounds like a workaround, stills forcing you to define a new "Copy without cancel" keybinding so you can reference it from your normal copy action, that includes canceling to exit.

Also I feel that having a fixed set of actions that automatically exits from the selection mode may not fit satisfy everyone, depending on which actions you consider that should getting out of there.

From my perspective, it sounds like a good alternative for this should be allowing multiple actions to be ran with a single keybinding, that would prevent the need of hacking the configuration. This would have the counterpart, however, that it would require you to think on every possible keystroke that you want to get you out of the selection mode, but that's probably more an issue of curating some default config that people can use to build on top of.

**Reply by doviethoa** [reply]
I recently switch to Zed from Emacs and also found not being able to deactivate selection mark mode automatically is a bit annoying. So I created #35677 to address it. The idea is to deactivate selection mark mode after an action by default unless it is any of the `select_<movement>` actions.

Could someone try it out and feedback please? Thank you.

**Comment by jinschoi** [comment]
I have noticed that with the new mark mode, C-space then Alt-< or Alt-> does not make a selection.

**Reply by notpeter** [reply]
@jinschoi What keyboard layout are you using?
When I type `ctrl-space` and then `alt-<` (requires `alt-shift-,` on my US keyboard) it selects from the current cursor to the beginning of the document.

Does normal `alt-<` work for you?

**Reply by jinschoi** [reply]
I had alt-< bound as MoveToBeginning in my custom keymap, from before it was supported. Removing it now binds it to SelectToBeginning, which works as expected. Never mind, and thanks.

**Comment by failable** [comment]
In Emacs, one can select and page up/down at the same time. I wish we can do it in Zed with and without the selection context. https://github.com/zed-industries/zed/issues/9015

**Reply by notpeter** [reply]
Can you explain further? If you want to be able to expand your selection to pages above and below the current selection point I think you can do something like: `ctrl-space pagedown ctrl-x ctrl-x pageup` -- are you looking for something different?

**Comment by akaihola** [comment]
@notpeter I opened PR #30858 to add support for `Ctrl-V` and `Alt-V` in selection mode.

**Comment by Velnbur** [comment]
Can Meta key be remapped to Command instead of Option in Mac?

**Reply by notpeter** [reply]
Not currently. You can do it in macOS for all apps, and I believe there are some third party apps which can do it per-application (BetterTouchTool perhaps?).

There's an issue. Upvote 👍 and subscribe to track progress:
- https://github.com/zed-industries/zed/issues/12708

**Comment by devcexx** [comment]
I've recently seen the new selection_mode feature integrated on Zed and it's really nice thing to have. It was actually one of my blockers to move to Zed so now I can start the transition! I've noted however, that the "SetMark" is likely missing a "PopMark" counterpart, emulating the Emacs mark ring. This is a feature similar to what is implemented for Vim (https://github.com/zed-industries/zed/pull/25702), but with a stack of past marks instead of having marks accessible with different keys.

**Reply by notpeter** [reply]
Would you be willing try and write up a scope for this feature request: the data structures, the actions/keybinds and their associated behaviors?

**Comment by failable** [comment]
Emacs key bindings like `Ctrl-x p f` have led to [unexpected behavior](https://github.com/zed-industries/zed/issues/32605) right now.

```
Zed: v0.200.1 (Zed Preview) 
OS: macOS 15.6.0
Memory: 64 GiB
Architecture: aarch64
```
